pkg/infra/httpclient/httpclientprovider/grafana_request_id_header_middleware.go

```diff
@@ -0,0 +1,59 @@
+package httpclientprovider
+
+import (
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/hex"
+	"net/http"
+
+	"github.com/google/uuid"
+
+	sdkhttpclient "github.com/grafana/grafana-plugin-sdk-go/backend/httpclient"
+	"github.com/grafana/grafana/pkg/services/pluginsintegration/clientmiddleware"
+	"github.com/grafana/grafana/pkg/setting"
+	"github.com/grafana/grafana/pkg/web"
+)
+
+const GrafanaRequestIDHeaderMiddlewareName = "grafana-request-id-header-middleware"
+
+func GrafanaRequestIDHeaderMiddleware(cfg *setting.Cfg) sdkhttpclient.Middleware {
+	return sdkhttpclient.NamedMiddlewareFunc(GrafanaRequestIDHeaderMiddlewareName, func(opts sdkhttpclient.Options, next http.RoundTripper) http.RoundTripper {
+		return sdkhttpclient.RoundTripperFunc(func(req *http.Request) (*http.Response, error) {
+			foundMatch := false
```

Comment:
The headers might already have been populated via https://github.com/grafana/grafana/blob/708aeb0682a7a7c74ec2d4ab6648880e0946bd27/pkg/services/pluginsintegration/clientmiddleware/grafana_request_id_header_middleware.go#L85-L86 
https://github.com/grafana/grafana/blob/708aeb0682a7a7c74ec2d4ab6648880e0946bd27/pkg/services/pluginsintegration/clientmiddleware/httpclient_middleware.go#L51-L58

So an optimization might be if headers already populated call `return next.RoundTrip(req)`

Why? In the case of a core plugin that handles a backend request and the plugin makes an outgoing HTTP request using a client from HTTP client provider it will first pass through client middleware and for the outgoing HTTP request httpclient middleware.

Reply:
Thanks for the explanation! I've added a check for it [here](https://github.com/grafana/grafana/pull/81662/files#diff-a9ab5a4883b435f5e7375904051ef744d351e22c8ca1a765b041de53b45b9b8cR17-R20).